fix(filetype): fix filetype patterns

smjonas opened the pull request:
Correct some incorrect filetype patterns and add missing ones to avoid test failures that became evident in #19216.

clason commented:
All the `starset` functions should be in the `pattern` table, not the `filename` table -- that should remove a sizeable chunk of errors.

clason commented on runtime/lua/vim/filetype.lua:
```diff
@@ -1305,7 +1304,7 @@ local filename = {
   ['/etc/apt/sources.list'] = 'debsources',
   ['denyhosts.conf'] = 'denyhosts',
-  ['.*/debian/patches/.*'] = function(path, bufnr)
+  ['*/debian/patches/.*'] = function(path, bufnr)
```
should be in `pattern` table

clason commented:
Another issue seems to be that Lua `.*` does _not_ match nothing, so, e.g. , `.*%.git/modules/.*/config` fails to match `file.git/modules//config`. @gpanders ?

smjonas commented:
What do you think about refactoring some patterns by omitting leading and trailing `.*` in Lua patterns?

> Another issue seems to be that Lua .* does not match nothing, so, e.g. , .*%.git/modules/.*/config fails to match file.git/modules//config.

This should be solved by using `.-` instead of `.*` (`.-` matches non-greedily).

clason commented:
> What do you think about refactoring some patterns by omitting leading and trailing `.*` in Lua patterns?

Hmmm... I think that would make it harder to check, for relatively little technical advantage.

> This should be solved by using `.-` instead of `.*` (`.-` matches non-greedily).

I don't think that's it, actually -- a "naked" `string.find` works, so this must be related to the `match_pattern` function (full path vs. tail)?

clason commented:
But in any case, the first order of business is to pass the test suite -- that will give us much more confidence when refactoring ;)

smjonas commented:
Now all tests are passing except the Git check (@gpanders):
```
1 FAILED:
Found errors in Test_filetype_detection():
function RunTheTest[37]..Test_filetype_detection[2]..CheckItems line 13: with file name: file.git/modules//config: Expected 'gitconfig' but got ''
make[1]: *** [Makefile:112: test_filetype_lua] Error 1
```

clason commented on runtime/filetype.lua:
```diff
@@ -13,7 +13,13 @@ vim.api.nvim_create_autocmd({ 'BufRead', 'BufNewFile' }, {
+      -- Generic configuration file used as fallback
+      ft = require('vim.filetype.detect').conf(vim.api.nvim_buf_get_name(args.buf), args.buf)
+      if ft then
+        vim.cmd('setf FALLBACK ' .. ft)
```
```suggestion
        vim.api.nvim_cmd({ cmd = 'setf', args = { 'FALLBACK', ft }, {})
```
more verbose but structured :)

clason commented on runtime/filetype.lua:
```diff
@@ -13,7 +13,13 @@ vim.api.nvim_create_autocmd({ 'BufRead', 'BufNewFile' }, {
-    if ft then
+    if not ft then
+      -- Generic configuration file used as fallback
+      ft = require('vim.filetype.detect').conf(vim.api.nvim_buf_get_name(args.buf), args.buf)
```
```suggestion
      ft = require('vim.filetype.detect').conf(args.file , args.buf)
```

clason commented:
For me also the `etc/hostname.file` check (expects `config`) fails?

clason commented on runtime/lua/vim/filetype.lua:
```diff
@@ -1693,6 +1690,9 @@ local pattern = {
   ['.*/%.?cmus/.*%.theme'] = 'cmusrc',
   ['.*/%.cmus/autosave'] = 'cmusrc',
   ['.*/%.cmus/command%-history'] = 'cmusrc',
+  ['/etc/hostname%..*'] = starsetf('config'),
```
```suggestion
  ['.*/etc/hostname%..*'] = starsetf('config'),
```
? (at least that passes the corresponding test for me.)

clason replied:
But that probably means the `match_pattern` function has an issue with its handling of `/` (which also relates to the `file.git/modules//config` failure).

clason replied:
That is interesting. Can you make sure you have pushed all your changes? Otherwise it'd be worrying that the tests behave differently on different platforms...

clason replied:
Definitely fails for me...

smjonas replied:
Yes, I pushed everything except the `finish` at the start of `filetype.vim`. 

> Otherwise it'd be worrying that the tests behave differently on different platforms...

I am on Linux Mint btw.

clason replied:
Ah, I think I see what's going on: this is indeed platform-specific: macOS sandboxes the `/etc` files; trying to edit `/etc/foo` will open `/private/etc/foo`. Hence Vim's pattern here will not work (and Vim in fact behaves the same). 

That means we have three options:
1. accept the failure (`/etc/hostname.*` is not relevant on macOS)
2. fix the failure (all other patterns are `*/etc/<foo>`, so this might be an oversight?)
3. report the issue upstream and see what they thing (@zeertzjq ?)

smjonas replied:
> report the issue upstream and see what they think

We should definitely do this one, maybe Vim will simply update the pattern to `*/etc/hostname.*` which we can then update as well

clason replied:
I'm quite sure it's an oversight, though -- _every single_ other of the >100 patterns uses `*/etc`.

smjonas replied:
Forgot to put a `finish` in `scripts.vim`, anything else? Because everything is failing right now.

clason replied:
Test fails on macOS CI as well; not sure yet why this is not an issue for `filetype.vim`, but that means we need to adapt the pattern to have the same behavior, independent of what Vim does.

smjonas replied:
Ok, will do that

clason commented:
> Now all tests are passing except the Git check (@gpanders):
> 
> ```
> 1 FAILED:
> Found errors in Test_filetype_detection():
> function RunTheTest[37]..Test_filetype_detection[2]..CheckItems line 13: with file name: file.git/modules//config: Expected 'gitconfig' but got ''
> make[1]: *** [Makefile:112: test_filetype_lua] Error 1
> ```

This seems to be something to do with actual filename normalization, since `vim.filetype.match({filename='file.git/modules//config'})` _does_ correctly return `'gitconfig'`.

On the other hand, `nvim file.git/modules//config` opens a buffer named `file.git/modules/config` (single slash). This is not the case for Vim.

So the two remaining problems have the same underlying issue: `filetype.vim` dispatches on `<amatch>`, which can be different from `<afile>` (which is basically equivalent to what `filetype.lua` is working with).

Three options:
1. refactor `filetype.lua` to use `<amatch>` instead of getting the filename
2. add filename manipulation so that our `path` matches exactly `<amatch>` (e.g., undoing some manipulations)
3. add new patterns to work around the specific test failures (and whichever get reported down the road).

My preference is option 3., not just because it's the simplest, but also because 1. doesn't work for using `vim.filetype.match` outside of the `FileType` autocommand and 2. would be a mess of special rules and not even fully possible.

In that case, I'd suggest
* adding the `.*` to the `/etc/hostname%..*` pattern
* adding a separate `file.git/modules/config` pattern

@gpanders ?

clason commented on runtime/filetype.lua:
```diff
@@ -13,7 +13,13 @@ vim.api.nvim_create_autocmd({ 'BufRead', 'BufNewFile' }, {
-    if ft then
+    if not ft then
+      -- Generic configuration file used as fallback
+      ft = require('vim.filetype.detect').conf(args.file, args.buf)
```
Based on my comment below about `<amatch>` vs. `<afile>`, this should probably be
```suggestion
      ft = require('vim.filetype.detect').conf(args.match, args.buf)
```

smjonas commented:
@clason I created a PR on the Vim repo that updates the `etc/hostname.*` pattern: https://github.com/vim/vim/pull/10662

clason commented:
I thought about this, but do we support filename and bufnr? I thought we didn't support combinations?

smjonas commented:
> but do we support filename and bufnr?

Yes we do since recently. 
(The original reason was to support checks like this:
```lua
bak = function(path, bufnr)
    local root = vim.fn.fnamemodify(path, ':r')
    return M.match({ buf = bufnr, filename = root })
end,
```
)

clason commented:
Yes, I think that sounds like a good idea.

smjonas commented:
Sure, do you want me to also include the "generic fallback" change in the autocommand in this separate commit or in the old one because it's more related to filetype changes?

clason commented:
> Sure, do you want me to also include the "generic fallback" change in the autocommand in this separate commit or in the old one because it's more related to filetype changes? And btw, is that a fix or refactor?

Good idea; the point is to separate the "infrastructure" changes that (may) have global effect from the changes to individual patterns.

I'd say it's a `fix`, not a `refactor`, as it has observable changes.

smjonas committed:
fix(filetype): fix filetype patterns

smjonas committed:
fix(filetype): remove call to vim.fn.resolve and pass filename to match function

For example on MacOS, /etc/hostname.file is symlinked to
/private/etc/hostname.file. We only care about the original file path though.

clason commented:
Thank you, @smjonas, for seeing this through!

smjonas commented:
Thank you as well for helping me with this! Seems like Lua filetype detection is almost ready, awesome! 🎉
